File: leetcode/go/697.degree-of-an-array.go
func findShortestSubArray(nums []int) int {
    left := make(map[int]int)
    right := make(map[int]int)
    counts := make(map[int]int)
    for i, n := range nums {
        if _, ok := left[n]; !ok {
            left[n] = i
        }
        right[n] = i
        counts[n]++
    }
    ans := len(nums)
    max_count := 0
    for k, v := range counts {
        if v > max_count {
            max_count = v
            ans = right[k] - left[k] + 1
        } else if v == max_count {
            ans = Min(ans, right[k]-left[k]+1)
        }
    }
    return ans
}

func Min(x int, y int) int {
    if x < y {return x}
    return y
}

File: leetcode/rust-leetcode/src/697.degree-of-an-array.rs
use std::collections::HashMap;
impl Solution {
    pub fn find_shortest_sub_array(nums: Vec<i32>) -> i32 {
        let mut map:HashMap<i32,(i32, (usize, usize))> = HashMap::new();
        let (mut max_n, mut max_count) = (vec![nums[0]], 1);
        for (i, &n) in nums.iter().enumerate() {
            match map.get_mut(&n) {
                Some(entry) => {
                    entry.0 += 1;
                    entry.1.1 = i;
                    if entry.0 > max_count {
                        max_n = vec![n];
                        max_count = entry.0;
                    } else if entry.0 == max_count {
                        max_n.push(n);
                    }
                },
                None => {
                    map.insert(n, (1, (i, i)));
                }
            }
        }
        let mut min_arr = i32::MAX;
        for n in max_n.iter() {
            match map.get(n) {
                None => unreachable!(),
                Some(entry) => {
                    min_arr = min_arr.min((entry.1.1 - entry.1.0 + 1) as i32);
                }
            }
        }
        min_arr
    }
}

